server/src/main/java/org/apache/iotdb/db/metadata/mtree/store/disk/cache/CacheMemoryManager.java

```diff
@@ -0,0 +1,125 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.metadata.mtree.store.disk.cache;
+
+import org.apache.iotdb.db.metadata.mtree.store.CachedMTreeStore;
+import org.apache.iotdb.db.metadata.mtree.store.disk.MTreeFlushTaskManager;
+import org.apache.iotdb.db.metadata.mtree.store.disk.MTreeReleaseTaskManager;
+import org.apache.iotdb.db.metadata.mtree.store.disk.memcontrol.IMemManager;
+import org.apache.iotdb.db.metadata.mtree.store.disk.memcontrol.MemManagerHolder;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class CacheMemoryManager {
+
+  private final List<CachedMTreeStore> storeList = new ArrayList<>();
+
+  private final IMemManager memManager = MemManagerHolder.getMemManagerInstance();
+
+  private final MTreeFlushTaskManager flushTaskManager = MTreeFlushTaskManager.getInstance();
+  private int flushCount = 0;
+  private volatile boolean hasFlushTask;
+
+  private final MTreeReleaseTaskManager releaseTaskManager = MTreeReleaseTaskManager.getInstance();
+  private volatile boolean hasReleaseTask;
+  private int releaseCount = 0;
+
+  public ICacheManager createLRUCacheManager(CachedMTreeStore store) {
+    ICacheManager cacheManager = new LRUCacheManager();
+    storeList.add(store);
+    return cacheManager;
+  }
+
+  public void ensureMemoryStatus() {
+    if (memManager.isExceedFlushThreshold() && !hasReleaseTask) {
+      registerReleaseTask();
+    }
+  }
+
+  private synchronized void registerReleaseTask() {
+    if (hasReleaseTask) {
+      return;
+    }
+    hasReleaseTask = true;
+    releaseTaskManager.submit(this::tryExecuteMemoryRelease);
+  }
+
+  /**
+   * Execute cache eviction until the memory status is under safe mode or no node could be evicted.
+   * If the memory status is still full, which means the nodes in memory are all volatile nodes, new
+   * added or updated, fire flush task.
+   */
+  private void tryExecuteMemoryRelease() {
+    for (CachedMTreeStore store : storeList) {
+      store.getLock().threadReadLock();
+      try {
+        store.executeMemoryRelease();
+      } finally {
+        store.getLock().threadReadUnlock();
+      }
+      if (!memManager.isExceedReleaseThreshold() || memManager.isEmpty()) {
+        break;
+      }
+    }
+    releaseCount++;
+    hasReleaseTask = false;
+    if (memManager.isExceedFlushThreshold() && !hasFlushTask) {
+      registerFlushTask();
+    }
+  }
+
+  private synchronized void registerFlushTask() {
+    if (hasFlushTask) {
+      return;
+    }
+    hasFlushTask = true;
+    flushTaskManager.submit(this::tryFlushVolatileNodes);
```

Comment:
Same problem as release task. Maybe these two boolean field should be attributes of store rather than a global unique attribute.

Reply:
I think the two global boolean variables are used to avoid duplicate release/flush requests during the release/flush process, which is needed since the release/flush thread is currently globally unique. When hasFlushTask is false, all store's hasFlushTask is also false, so there seems to be no need to migrate to store?

Reply:
Use CompleteFuture to release and flush concurrently.